modules/apps/commerce/headless/headless-commerce/headless-commerce-admin-shipment-impl/src/main/java/com/liferay/headless/commerce/admin/shipment/internal/util/v1_0/ShipmentItemUtil.java

```diff
@@ -52,16 +59,91 @@ public static CommerceShipmentItem addOrUpdateShipmentItem(
 				commerceShipmentItem.getCommerceInventoryWarehouseId();
 		}
 
+		CommerceInventoryWarehouse commerceInventoryWarehouse =
+			getCommerceInventoryWarehouse(
+				serviceContext.getCompanyId(), shipmentItem,
+				commerceInventoryWarehouseService);
+
+		CommerceOrderItem commerceOrderItem = getCommerceOrderItem(
+			serviceContext.getCompanyId(), shipmentItem,
+			commerceOrderItemService);
+
 		return commerceShipmentItemService.addOrUpdateCommerceShipmentItem(
 			externalReferenceCode, commerceShipment.getCommerceShipmentId(),
 			GetterUtil.getLong(
-				shipmentItem.getOrderItemId(), defaultOrderItemId),
+				commerceOrderItem.getCommerceOrderItemId(), defaultOrderItemId),
 			GetterUtil.getLong(
-				shipmentItem.getWarehouseId(), defaultWarehouseId),
+				commerceInventoryWarehouse.getCommerceInventoryWarehouseId(),
+				defaultWarehouseId),
 			BigDecimalUtil.get(shipmentItem.getQuantity(), defaultQuantity),
 			null,
 			GetterUtil.getBoolean(shipmentItem.getValidateInventory(), true),
 			serviceContext);
 	}
 
+	public static CommerceInventoryWarehouse getCommerceInventoryWarehouse(
+			long companyId, ShipmentItem shipmentItem,
+			CommerceInventoryWarehouseService commerceInventoryWarehouseService)
+		throws Exception {
+
+		long commerceInventoryWarehouseId = GetterUtil.getLong(
+			shipmentItem.getWarehouseId());
+
+		CommerceInventoryWarehouse commerceInventoryWarehouse = null;
+
+		if (commerceInventoryWarehouseId > 0) {
+			commerceInventoryWarehouse =
+				commerceInventoryWarehouseService.getCommerceInventoryWarehouse(
+					commerceInventoryWarehouseId);
+		}
+		else {
+			String warehouseExternalReferenceCode =
+				shipmentItem.getWarehouseExternalReferenceCode();
+
+			commerceInventoryWarehouse =
+				commerceInventoryWarehouseService.fetchByExternalReferenceCode(
+					warehouseExternalReferenceCode, companyId);
+
+			if (commerceInventoryWarehouse == null) {
+				throw new NoSuchInventoryWarehouseException(
+					"Unable to find warehouse with external reference code " +
+						warehouseExternalReferenceCode);
+			}
+		}
+
+		return commerceInventoryWarehouse;
+	}
+
+	public static CommerceOrderItem getCommerceOrderItem(
+			long companyId, ShipmentItem shipmentItem,
+			CommerceOrderItemService commerceOrderItemService)
+		throws Exception {
+
+		long commerceOrderId = GetterUtil.getLong(
+			shipmentItem.getOrderItemId());
+
+		CommerceOrderItem commerceOrderItem = null;
```

Comment:
I think this should go on top

Reply:
I think this method should return only the id (is it possible or we use other params)? and the logic should be
```
if (commerceOrderItemId > 0) return commerceOrderItemId;
commerceOrderItem =
				commerceOrderItemService.fetchByExternalReferenceCode(
					orderItemExternalReferenceCode, companyId);
					if (commerceOrderItem == null) return 0;
return commerceOrderItem.getCommerceOrderItemId();
```

Reply:
same for warehouses

Reply:
and in case we return only ids change also the method names